server/src/main/java/org/elasticsearch/index/mapper/SourceFieldMapper.java

```diff
@@ -240,8 +243,17 @@ protected void parseCreateField(ParseContext context, List<IndexableField> field
             }
             BytesRef ref = source.toBytesRef();
             fields.add(new StoredField(fieldType().name(), ref.bytes, ref.offset, ref.length));
+        } else {
+            source = null;
         }
-    }
+
+        if (originalSource != null && source != originalSource && context.indexSettings().isSoftDeleteEnabled()) {
+            // if we omitted source or modified it we add the _recovery_source to ensure we have it for ops based recovery
+            BytesRef ref = source.toBytesRef();
```

Comment:
Should we store the original source here?

Reply:
LOL yeah we should :D - I will fix